accountability/accountability.go

```diff
@@ -1,45 +1,58 @@
+	"github.com/mikanikos/Fork-Accountability/common"
 )
 
-// faulty set
-var faultyProcesses *FaultySet
+// Accountability stores all the validators that are faulty and the corresponding faultiness proofs
+type Accountability struct {
+	HeightLogs *HeightLogs
+	FaultySet  *FaultySet
+}
+
+// NewAccountability creates a new Accountability structure
+func NewAccountability() *Accountability {
+	return &Accountability{
+		HeightLogs: NewHeightLogs(),
+		FaultySet:  NewFaultySet(),
+	}
+}
+
+// String returns a string representation (result) of the accountability algorithm
+func (acc *Accountability) String() string {
+	var sb strings.Builder
+	sb.WriteString("Accountability algorithm overview: \n")
+	sb.WriteString(acc.HeightLogs.String())
+	sb.WriteString(acc.FaultySet.String())
+	return sb.String()
+}
 
 // IdentifyFaultyProcesses detects which processes caused the fork and finds all processes that have bad behavior
-func IdentifyFaultyProcesses(numProcesses, firstDecisionRound, secondDecisionRound uint64, hvsMap *common.HeightLogs) *FaultySet {
+func (acc *Accountability) IdentifyFaultyProcesses(numProcesses, firstDecisionRound, secondDecisionRound uint64) {
 
-	faultyProcesses = NewFaultySet()
+	// lock logs to prevent other additions during the execution
+	acc.HeightLogs.mutex.Lock()
+	defer acc.HeightLogs.mutex.Unlock()
 
 	// first, preprocess messages by scanning all the received vote sets and add missing messages in the processes which omitted sent messages
-	preprocessMessages(numProcesses, firstDecisionRound, secondDecisionRound, hvsMap)
-
-	// then, find faulty processes by analyzing the message logs
-	findFaultyProcesses(numProcesses, firstDecisionRound, secondDecisionRound, hvsMap)
+	acc.preprocessMessages(numProcesses, firstDecisionRound, secondDecisionRound)
 
-	return faultyProcesses
+	// then, find faulty processes by analyzing the message HeightLogs
+	acc.findFaultyProcesses(numProcesses, firstDecisionRound, secondDecisionRound)
 }
 
 // Preprocess messages by scanning all the received vote sets and add missing messages in the respective votes sets of processes which omitted sent messages
-func preprocessMessages(numProcesses, firstDecisionRound, secondDecisionRound uint64, hvsMap *common.HeightLogs) {
+func (acc *Accountability) preprocessMessages(numProcesses, firstDecisionRound, secondDecisionRound uint64) {
 	for round := firstDecisionRound; round <= secondDecisionRound; round++ {
 		for processIndex := uint64(1); processIndex <= numProcesses; processIndex++ {
 
-			hvs, hvsLoad := hvsMap.Logs[processIndex]
-			// if process didn't send the hvs, it's faulty
+			hvs, hvsLoad := acc.HeightLogs.logs[processIndex]
+			// if process didn't send the hvs, just ignore
 			if hvs == nil || !hvsLoad {
-				faultyProcesses.AddFaultinessReason(NewFaultiness(processIndex, 0, ErrHVSNotSent))
+				//acc.FaultySet.AddFaultinessReason(NewFaultiness(processIndex, 0, FaultinessHVSNotSent))
```

Comment:
Even if monitor did not receive the Height-Vote-Set of some process P, it must execute this.
Why?
Because, monitor may notice equivocation and will be able to expose some processes.

Let me give you an example.
Suppose that you have that two correct process P and Q decide differently in round R.
Hence, there is an equivocation.
Moreover, suppose that monitor receives Height-Vote-Sets from f + 1 correct guys (hence, it does not receive Height-Vote-Sets of faulty processes that committed equivocation).
By ignoring them (since monitor did not receive their Height-Vote-Sets), monitor will not be able to expose them (and note that these faulty guys will never send their Height-Vote-Sets to monitor).

Please, let me know if you cover this scenario, but I did not understand it.

Reply:
That's perfectly true, I removed that line purposely because I thought that the new async version couldn't handle some cases. But I'm still not sure if adding that line would solve all problems. For instance, what happens in the opposite scenario?

Imagine the same example you described but the monitor receives Height-Vote-Sets from f + 1 faulty guys first (correct validators are slow). The monitor runs the algorithm because it received f+1 message logs: since A and B didn't send their logs, the monitor will detect them and the algorithm completes by wrongly stating that A and B are faulty.

It's likely that I'm wrong and I'm not considering something important, so let me know what I'm missing. By the way, I'm mostly relying on the document you sent me by email, I don't know if you updated it or something changed.

Reply:
I am not sure I understand which processes A and B are?
Note the following, we assume that messages can not be forged, i.e., some process P can claim that it has received message M from process Q only if Q indeed sent this message.

Reply:
Sorry I modified my previous comment: I was writing a different example, then changed my mind but forgot to update the last part. What I'm saying is that the monitor can wrongly detect some processes only because they have been slow to send their logs: since the algorithm is run once the monitor has received f+1 message logs, if we detect all the other processes that didn't send their logs yet the algorithm would not work correctly.

Reply:
Not really, that's the beauty of it.
Monitor will run it for the first time once it delivers f + 1 Height-Vote-Sets and will try to expose f + 1 processes. If monitor does not succeed, it will wait for new Height-Vote-Sets.

The key here is to understand the following: Once monitor receives Height-Vote-Sets from all correct processes, it will be able to expose at least f + 1 faulty processes.
Hence, if monitor does not succeed in exposing f + 1 processes in the first "run", it can be sure that it will receive more Height-Vote-Sets.
Did I clarify things now?

Reply:
Yes I know that the monitor will wait for new HVSs if it fails detecting f+1 processes but, in this case, it will expose f+1 processes just because they didn't send their HVSs. I'll give you the example I was writing previously which maybe it's more clear:

Suppose we have 4 processes, two are correct (A and B) and two are faulty (C and D).
The monitor receives the HVSs from C and D (assume A and B are slow): since the monitor received f+1 HVSs, it runs the algorithm and it will detect A and B by executing that line above, just because it didn't receive the HVSs of A and B yet. Since the monitor exposed f+1 processes (A and B), the monitor stops. Please let me know if it's more clear now.

Reply:
Yes, but you did not understand my comment.

So, monitor will not expose them because they did not send their Height-Vote-Set. It will just add messages C and D received from A or B to Height-Vote-Sets of A and B (that are empty since monitor did not receive them).
Then, monitor will run through Height-Vote-Sets of A and try to find equivocation and it will not succeed since A is correct (same for B).
Bottom line is, monitor does not expose some process P if it did not receive its Height-Vote-Set, but only if it figures that P equivocated.
You see the difference?

Reply:
You're right, I see the difference now! However, adding the above line would not be completely correct, it must be added when we check if a process is faulty and not during the pre-processing phase. Thanks a lot for that, I'll fix this.

Reply:
To better explain the problem to whoever is gonna read this, adding the above line would be wrong because we are just ignoring processes from the pre-processing phase, not from the faultiness-detection algorithm. There's just one small bug few lines later that must be fixed because the monitor should initialize empty Height-Vote-Sets even if it didn't receive anything from some processes.

Reply:
Actually, the last fix we discussed yesterday during the call is not enough and generated another problem. We can discuss it later in the scheduled call.

Reply:
Can you briefly explain the problem here?
Thanks :)

Reply:
As we know, the implemented algorithm has two consecutive phases: pre-processing phase and fault-detection phase.

In the pre-processing phase, we go through all the received height vote sets so far and we "infer" the missing height vote sets of the processes that didn't send their logs yet. However, during this phase, we only infer the sent messages by the processes that didn't send their logs based on the received height vote sets that the monitor collected.

When it comes to the second phase, we run the algorithm to check whether a process is faulty or not (based on its height vote set, either received or inferred) but we cannot use the inferred height vote sets to check for a faultiness different than equivocation. 

In fact, when checking if a process received enough (2f+1) prevotes to send a precommit (or enough prevotes to send another prevote in a later round), the process with the inferred height vote set will not contain any message in its received messages bucket because the pre-processing phase infers only the messages that have been sent by the process. Therefore, the process will always result faulty because it doesn't have enough prevotes to justify a sent precommit (or prevote) since it doesn't have any received messages in any round.

As far as I understood, the inferred height vote sets should only be checked for equivocation (i.e. a process sent two or more messages with the same type in the same round but with different value). Would this assumption be correct?

Please let me know if everything is clear.

Reply:
I understood, thanks! :)
However, even in these cases the monitor will make a correct call (that is why for asynchronous cases some messages need to contain justification).
So, the fix we had yesterday is ok.

I will explain in more details everything today.
Please, do not change the code before our meeting, so we can discuss this scenario,
Thanks!

Reply:
I've already changed and refactored the code yesterday but we can start from this commit https://github.com/mikanikos/Fork-Accountability/tree/e3327e0fd55e40156fff30dfdbac068d6fa05585.

Reply:
As we said, inferred height vote sets should only be checked for equivocation. Fixed in the last commit.